Protect against symbols with unsafe chars

RuijieYu opened the pull request:
Fixes #648.  Also fixes #653 by its first commit.

* lispy.el lispy--symbol-safe-chars-regexp: define safe chars.
(lispy--read): protect against unsafe chars in a symbol.

@simendsjo: Can you check if this works?

RuijieYu committed:
Protect against symbols with unsafe chars

Fixes bug#648.

* lispy.el lispy--symbol-safe-chars: define list of safe chars.
(lispy--read): protect against unsafe chars in a symbol.
(lispy--replace-regexp-in-code): add predicate to only replace on match data.

simendsjo commented:
> @simendsjo: Can you check if this works?

Almost. `(.> foo)` no longer escapes, but `(.)` actually removes all three characters.

RuijieYu commented:
This iteration should fix the immediate problem (and I have some commits coming after for cleaning up unnecessary code).  Let me know if it works.  The culprit was that `lispy--bounds-dwim` wasn't able to get its bounds, and now it should.

simendsjo commented:
>  Let me know if it works.

Not sure if I'm doing anything strange or if it's something Doom related which causes problems, but when testing 420363867dad42d0d0ad6ffa141524deae7b91ea, pretty much nothing works. Pressing `(`, I get `()`, then `d` to get `(d)`, then trying to add `e` at the following where `|` is the cursor `(d|)` gives `forward-list: Scan error: "Containing expression ends prematurely", 1, 1`

RuijieYu commented:
Got it. Looks like I misunderstood what `lispy-right` represents.

RuijieYu committed:
Fixed test file, and added regression test cases

Fixes bug#653.

* lispy-test.el
(lispy-simulate-key) (lispy-simulate-keys) (lispy-simulate-expect):
Added facilities to assert results after a given set of keypresses.
(lispy-read-unsafe-chars): added regression tests for bug#648.

RuijieYu commented:
Does the current iteration work?  I wrote some tests incorporating what you have tested so far, and it seems that the tests have passed on my end.

simendsjo commented:
> Does the current iteration work? I wrote some tests incorporating what you have tested so far, and it seems that the tests have passed on my end.

Yes, looks like it's working! Thanks!

RuijieYu commented:
Blocked by abo-abo/lispy#658.  Don’t have much time ATM.
